src/app/linux.rs

```diff
@@ -1,20 +1,449 @@
+    fmt,
+    fs,
+    io,
+    path::{Path, PathBuf},
+};
 
 use super::Snapshot;
 
-pub(super) struct SnapshotIter {}
+/// SnapshotIter allocates a vector of the paths to objsets
+/// under the desired pools on initialization.  Reading the
+/// objsets into Snapshots is lazy.
+pub(super) struct SnapshotIter {
+    idx:     usize,
+    objsets: Vec<PathBuf>,
+}
+
+/// This is the default filepath for zfs stats using ZoL.
+const DEFAULT_ZFS_STATS_PATH: &str = "/proc/spl/kstat/zfs";
 
 impl SnapshotIter {
-    pub(crate) fn new(_pool: Option<&str>) -> Result<Self, Box<dyn Error>> {
-        unimplemented!()
+    /// Simply use the default path to the stats files in procfs.  A value of `None`
+    /// for `pool` will get statistics for all pools.
+    pub(crate) fn new(pool: Option<&str>) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_basepath(DEFAULT_ZFS_STATS_PATH, pool)
+    }
+
+    // Useful for testing on a mock directory.
+    fn new_from_basepath(
+        basepath: &str,
+        pool: Option<&str>,
+    ) -> Result<SnapshotIter, Box<dyn Error>> {
+        let objsets = Self::required_pools_from_basepath(basepath, pool)?
+            .into_iter()
+            .flat_map(Self::enumerate_pool)
+            .collect::<Vec<PathBuf>>();
+        Ok(SnapshotIter { idx: 0, objsets })
+    }
+
+    // Get all objset paths for a given pool.
+    // At this point all the paths should exist, because they have been
+    // checked by `required_pools_from_basepath`.  Should this somehow be
+    // expressed as an invariant?
+    fn enumerate_pool(pool: PathBuf) -> Vec<PathBuf> {
+        let is_dataset = |entry: &fs::DirEntry| {
+            let path = entry.path();
+            let name = entry.file_name();
+            let name = name.to_str().unwrap_or("");
+            path.is_file() && name.starts_with("objset")
+        };
+        fs::read_dir(pool).map_or(vec![], |dir| {
+            dir.filter_map(|e| {
+                e.map_or(None, |entry| {
+                    if is_dataset(&entry) {
+                        Some(entry.path())
+                    } else {
+                        None
+                    }
+                })
+            })
+            .collect()
+        })
+    }
+
+    fn required_pools_from_basepath(
+        basepath: &str,
+        pool: Option<&str>,
+    ) -> Result<HashSet<PathBuf>, Box<dyn Error>> {
+        let zfs_stats_path = PathBuf::from(basepath);
+        let mut pools = SnapshotIter::get_pools(zfs_stats_path.as_path())
+            .map_or_else(|err| Err(Box::new(err)), Ok)?;
+        if let Some(pool) = pool {
+            let pool: PathBuf =
+                [zfs_stats_path, PathBuf::from(pool)].iter().collect();
+            pools.retain(|p| p == &pool);
+            if pools.is_empty() {
+                let name = pool
+                    .file_name()
+                    .map(|ostr| ostr.to_str().unwrap_or(""))
+                    .unwrap_or("")
+                    .to_string();
```

Comment:
```suggestion
                let name = pool.to_owned();
```
For this to work, you need to not shadow the `pool` binding on line 76.

Reply:
Good idea.  I implemented it.